### ISSUE TITLE
Design discussion:  Resolution of fields of value type for {get|put|with}field operations in JIT compiled code

### ISSUE THREAD

**Issue by hzongaro** [eclipse/openj9#11729]
The following summarizes some off-line discussion between Leonardo @Leonardo2718, Annabelle @a7ehuo, Daryl @0xdaryl and myself.

In the prototype implementation of value types, a field whose type is a value type might be flattened by the JVM.  The contents of such a field will be held directly in the containing object.

For resolution of a field whose type is not a value type, the IL will contain a `ResolveCHK` operation whose child is a memory reference with the appropriate form.  For example, for the reference to `o.v`, in the following

```
public class Holder {Object v;}
public class Toy {
   public static Object sub(Holder o) {return (o == null) ? null : o.v;}
}
```

the IL will look like this

```
n18n   ResolveCHK [#32]
n17n     aloadi  LHolder.v Ljava/lang/Object;
n16n       aload  <parm 0 LHolder;>  // Load of "o"
```

In the case of an unresolved field of value type, as the field might be flattened in the containing object, the JIT will not be able to construct a simple memory reference for the field.  If the field and its type have been resolved, the JIT is able to inquire of the JVM how the contents of that field are stored in an instance of the containing class, and will generate loads or stores of the individuals parts of the flattened field of value type.  Alternatively, the JIT is also able to generate calls to JVM helpers to retrieve the value from the flattened field or to store the value to the flattened field.  Which is chosen is determined by the settings of the `useFlattenedFieldRuntimeHelpers` and `dontUseFlattenedFieldRuntimeHelpers` JIT options.

In the following example, if the `v` field of `Holder` is flattened, and the JIT is generating calls to the JVM flattened field helpers

```
public value Point {double x, y;}
public class Holder {Point v;}
public class Toy {
   public static Object sub(Holder o) {return (o == null) ? null : o.v;}
}
```

the IL generated by the JIT looks like this

```
n21n   treetop
n20n     acall  jitGetFlattenableField
n16n       aload  <parm 0 LHolder;>   // Load of "o"
n17n       aconst 0x17e210   // cp()+cpIndex
```

The JVM team will provide a runtime helper, `jitResolveFlattenableField`, that the JIT can use to perform the resolution of a field whose type is a value type, as described in #11570

> ValueTypes: Add jit runtime helper to resolve field refs for flattenable fields:  
> 
>   Params:
>     J9VMThread currentThread: implicit param
>     J9Method method: method in which the resolution must take place
>     U_32 cpIndex: ram constant pool index for the field ref
>     U_32 resolveType: one of J9TR_FLAT_RESOLVE_WITHFIELD / J9TR_FLAT_RESOLVE_GETFIELD / J9TR_FLAT_RESOLVE_PUTFIELD
> 
>     For U_32 parameters on 64-bit platforms, the high-order 32 bits are unused.

The JVM helpers `jit{Get|Put|With}FlattenableField` are able to handle fields whether they are actually flattened or not, so the JIT can generate a call to use one of those helpers for any unresolved field that is of a value type, despite not having information about the actual layout of the field.  So, the idea would be for the JIT to generate code for a call to a snippet that would call `jitResolveFlattenableField` to perform the resolution, and then patch the mainline call to the snippet with a `nop`, so that the resolution would likely only be performed once at that point in the code.  That snippet call would be followed by the call to `jit{Get|Put|With}FlattenableField`.

The question to be resolved in this issue is how to represent the required resolve check operation in the IL, as there would be no explicit memory reference in the IL with which to associate the resolve check of the field - just a helper call.  We discussed two possibilities:

1. Using a call to a non-helper symbol, say `valueTypeFieldResolveCHK`, whose arguments would be the `cpIndex` and the kind of operation
```
    n99n   treetop
    n98n     acall  valueTypeFieldResolveCHK
    n97n       iconst 5  // cpIndex
    n96n       iconst 1  // J9TR_FLAT_RESOLVE_GETFIELD
    n21n   treetop
    n20n     acall  jitGetFlattenableField
    n16n       aload  <parm 0 LHolder;>
    n17n       aconst 0x17e210   // cp()+cpIndex
 ```
     
2. Defining a new ValueTypeFieldResolveCHK opcode with operands being the `cpIndex` and the kind of operation involved:
```
    n99n   ValueTypeFieldResolveCHK
    n98n     iconst 5  // cpIndex
    n97n     iconst 1  // J9TR_FLAT_RESOLVE_GETFIELD
    n21n   treetop
    n20n     acall  jitGetFlattenableField
    n16n       aload  <parm 0 LHolder;>
    n17n       aconst 0x17e210   // cp()+cpIndex
 ```

**Comment by hzongaro** [comment]
Vijay @vijaysun-omr, FYI - we're hoping to get your thoughts on the best way to represent this in the IL.

**Comment by vijaysun-omr** [comment]
Thanks @hzongaro for the writeup. 

Have you considered a resolve check tree structure similar to the unresolved load/store for non-value types ? i.e. a resolve check above the load or store opcode for the unresolved field ? I ask, because at some level it does express the "true" semantics of what is going on, namely that we are loading or store a field with a particular signature (`Holder.v` in your example). This fact gets obfuscated to some degree with the helper calls being used in the IL and of course you also have the problem that you referred to in the issue, namely how to represent the resolve check.

I do understand that representing it as a helper call in the IL makes the codegen-specific changes minimal and so it saves development effort in this case but I wonder if we also end up adding enough quirks like what you are proposing above to make us at least reconsider if in fact the call representation is the best one *in this case*.  Taking unresolved accesses and generating a call from them in the evaluators is a fairly common operation that we seem to have done before for non-value type cases and it may be that you just need to call a different helper, but otherwise you can reuse a lot of the codegen logic that is already there (and well tested). 

I'd also suggest maybe not considering a patch of the mainline code for such unresolved cases at least, i.e. call a helper and just live with that overhead every time on the assumption that it will almost never run in practice and it eases the implementation effort significantly to not have to worry about code patching.

**Comment by vijaysun-omr** [comment]
I suppose another point on the design spectrum could be to lower the unresolved load/store of a value type field into the kind of call that you showed above but do that in `lowerTrees` so as to minimize codegen-specific changes. Note that the resolve check does not change the generated code for the unresolved access, it merely represents the fact that an exception is possible during the resolution step (and any exception edges that emanate at that block cannot be removed if the catcher can be the handling block for the exception) to the optimizer. So, if you lowered it in `lowerTrees` it might be okay to not even have the resolve check at all since the optimizer would have finished running by then.

**Comment by Leonardo2718** [comment]
> Have you considered a resolve check tree structure similar to the unresolved load/store for non-value types ? i.e. a resolve check above the load or store opcode for the unresolved field ? I ask, because at some level it does express the "true" semantics of what is going on, namely that we are loading or store a field with a particular signature (Holder.v in your example). This fact gets obfuscated to some degree with the helper calls being used in the IL and of course you also have the problem that you referred to in the issue, namely how to represent the resolve check.

The concern is that the helper call (i.e. the load/store) does not contain the information that needs to be passed to the resolution helper, so we need to use something else to pass that information along. We also can't use an actual load/store opcode because the exact sequence of trees depends on the layout of the type, which is only known after the field is resolved (a load of a flattened value type looks like an allocation followed by a sequence of writes of each nested flattened field).

> I'd also suggest maybe not considering a patch of the mainline code for such unresolved cases at least, i.e. call a helper and just live with that overhead every time on the assumption that it will almost never run in practice and it eases the implementation effort significantly to not have to worry about code patching.

Yeah, that's a good point. We're including this more as a possibility than a definit requirement.

> I suppose another point on the design spectrum (related to my prior suggestion) could be to lower the unresolved load/store of a value type field into the kind of call that you showed above but do that in lowerTrees so as to minimize codegen-specific changes. 

I'm not entirely shure how this would work. We would still need to define something that captures the information needed for the resolution helper call. I guess we could define a non-helper symbol (or new opcode, as appropriate) that represents the combined resolution+memory operation. I _believe_ there is also a restriction with the current implementation of the resolution helper where it must be called from an unresolved data snippit. Not sure yet if that is a problem for doing the work in lowerTrees.

**Comment by vijaysun-omr** [comment]
> The concern is that the helper call (i.e. the load/store) does not contain the information that needs to be passed to the resolution helper, so we need to use something else to pass that information along. 

Could you please elaborate on that ? Is it something other than the constant pool entry (since that can conveyed on the load or store) ?

**Comment by vijaysun-omr** [comment]
> We also can't use an actual load/store opcode because the exact sequence of trees depends on the layout of the type, which is only known after the field is resolved (a load of a flattened value type looks like an allocation followed by a sequence of writes of each nested flattened field).

If you lowered to the helper call, would'nt that call be the one responsible for doing the allocation followed by a sequence of each nested flattened field ? i.e. all these details would not need to be exposed explicitly in the IL trees and all you would have is the load/store to begin with and the helper call after the lowering ? Just checking if that would not work for some reason...

**Comment by hzongaro** [comment]
Thanks for feedback, Vijay and Leo.

> > We also can't use an actual load/store opcode because the exact sequence of trees depends on the layout of the type, which is only known after the field is resolved (a load of a flattened value type looks like an allocation followed by a sequence of writes of each nested flattened field).

> If you lowered to the helper call, would'nt that call be the one responsible for doing the allocation followed by a sequence of each nested flattened field ? i.e. all these details would not need to be exposed explicitly in the IL trees and all you would have is the load/store to begin with and the helper call after the lowering ? Just checking if that would not work for some reason...

Yes, I think you are correct - that approach could work in the cases of `putfield` and `getfield`, but I'm not sure whether that would work for `withfield`, which still would not have a memory reference associated with it.  I didn't emphasize the case of `withfield` sufficiently in the introductory comments.

I will draw up some examples of the IL that's generated for `withfield` in the current prototype implementation.

**Comment by hzongaro** [comment]
I had promised some examples of the IL that's generated for `withfield`.

Consider the following pseudo-code.  `Point` is a value type class with two fields of type `double`, and `Line` is a value type class with two fields of type `Point`.  Some method `sub` applies a `withfield` operation to the value of type `Line`, which results in another value of type `Line` with the same value for the `p2` field as the original `Line` value, but the value of the `p1` field equal to another `Point` value that appeared as the second argument to `sub`.

```
value Point {double x; double y;}
value Line {Point p1; Point p2;}
public static Object sub(Object o, Object v) {
    return (o == null) ? null : (((Line) o) with { p1 = (Point) v; });
}
```

The bytecode for `(((Line) o) with { p1 = (Point) v; })` would look like this:

```
       8: aload_0
       9: checkcast     #2                  // class Unres4$Line
      12: aload_1
      13: checkcast     #3                  // class Unres4$Point
      16: withfield     #62                 // Field Unres4$Line.p1:QUnres4$Point;
```

If the field reference for `Line.p1` is already resolved when the JIT is compiling `sub`, and it **is not** a flattened field, the IL generated by the JIT looks like the following.  It creates a new `Line` instance by using a `newvalue` instruction with operands being the address of the `Line` class, the result of `(Point) v`, and a reference to the existing `((Line)o).p2` field.  It's not until the `newvalue` operation is expanded in `lowerTrees` that a memory reference for a `Line.p1` field is actually created.

```
      n18n    checkcast
      n16n      aload  <parm 0 Ljava/lang/Object;>
      n17n      loadaddr  Unres4$Line
      n23n    NULLCHK on n16n
      n22n      PassThrough
      n16n        ==>aload
      n26n    NULLCHK on n16n
      n25n      aloadi  Unres4$Line.p2 QUnres4$Point;
      n16n        ==>aload
      n28n    compressedRefs
      n25n      ==>aloadi
      n27n      lconst 0
      n30n    treetop
      n29n      newvalue  jitNewValue
      n24n        loadaddr  Unres4$Line
      n19n        ==>aload
      n25n        ==>aloadi
```

If the field reference for `Line.p1` is already resolved when the JIT is compiling `sub`, and **is** a flattened field, and the JIT is using the JVM's flattened field helpers, the IL generated by the JIT looks like the following - it uses a call to the `jitWithFlattenableField` with the original `Line` object as the first operand, the new `Point` value for the `p1` field as the second operand, and `cp()+cpIndex` encoded as the third operand.

```
      n18n    checkcast 
      n16n      aload  <parm 0 Ljava/lang/Object;>
      n17n      loadaddr  Unres4$Line
      n21n    checkcast 
      n19n      aload  <parm 1 Ljava/lang/Object;>
      n20n      loadaddr  Unres4$Point
      n23n    NULLCHK on n16n 
      n22n      PassThrough                                                                        
      n16n        ==>aload
      n26n    treetop                                                                              
      n25n      acall  jitWithFlattenableField
      n19n        ==>aload
      n16n        ==>aload
      n24n        aconst 0x186250  
```

In both cases, there's no memory reference for `Line.p1` with which to associate a `ResolveCHK`.

**Comment by vijaysun-omr** [comment]
That is a very interesting example @hzongaro with the withfield bytecode.

For the "not flattended" case my initial thought is to allow a resolve check above the newvalue opcode (since the newvalue opcode is a recent addition to the IL I expect us to have the freedom to define these kinds of IL shape/relationship properties still without worrying too greatly about making it work with existing code). The newvalue opcode is at least implicitly akin to a store (except that we don't want to expose the store to the optimizer lest it be be separated from the associated new allocation) and so in a sense we could allow a resolve check above it in an analogous manner to what we do for stores. We could in theory put the symbol reference for the field that was mentioned in the withfield bytecode on the newvalue opcode since I believe (rightly ?) that we do not really need the existing symbol reference on newvalue in any way during the optimizer especially since it gets lowered into new eventually anyway.

For the "flattened" case, I wonder if the same newvalue opcode (with the above proposed change to use the resolve check above it and the symbol reference for the field mentioned in the withfield bytecode) could be used at least through the optimizer and then only lower it differently before codegen into that jitWithFlattenableField helper if you wanted to do so. 

This has one other advantage of keeping the IL trees that the optimizer sees relatively consistent in these cases and the differences between flattened and not flattened only occur during lowering.

**Comment by hzongaro** [comment]
> Have you considered a resolve check tree structure similar to the unresolved load/store for non-value types ? i.e. a resolve check above the load or store opcode for the unresolved field ?

Just to summarize what I understand Vijay @vijaysun-omr is suggesting, for the following example

```
public value Point {double x, y;}
public class Holder {Point v;}
public class Toy {
   public static Object sub(Holder o) {return (o == null) ? null : o.v;}
}
```

the JIT could generate the following IL, where `LHolder.v` would be a fake shadow symbol:

```
    n99n   ResolveCHK [#32]
    n98n     aloadi  LHolder.v
    n97n       aload  <parm 0 LHolder;>  // Load of "o"
    n21n   treetop
    n20n     acall  jitGetFlattenableField
    n97n       ==>aload
    n17n       aconst 0x17e210   // cp()+cpIndex
```

Similarly for an assignment to `o.v`, it could generate the following IL:

```
    n99n  ResolveCHK [#32]
    n98n    astorei LHolder.v
    n97n      aload  <parm 0 LHolder;>  // Load of "o"
    n96n      aload  <parm 1 QPoint;>    // Load of parm of type Point
    n21n   treetop
    n20n     acall  jitPutFlattenableField
    n96n       ==>aload
    n97n       ==>aload
    n17n       aconst 0x17e210   // cp()+cpIndex
```

Further, the JIT could translate the `ResolveCHK` to the helper call during code generation or during lowerTrees, with the appropriate value for the `resolveType` argument:  `J9TR_FLAT_RESOLVE_GETFIELD` or `J9TR_FLAT_RESOLVE_PUTFIELD`.

And then in the case of `withfield`, for the following code:

```
value Point {double x; double y;}
value Line {Point p1; Point p2;}
public static Object sub(Object o, Object v) {
    return (o == null) ? null : (((Line) o) with { p1 = (Point) v; });
}
```

the JIT might generate the following IL:

```
      n30n    ResolveCHK [#32]
      n29n      newvalue  LUnres4$Line.p1
      n24n        loadaddr  Unres4$Line
      n19n        ==>aload
      n25n        ==>aloadi
      n99n    astore <auto slot 2>
      n29n      ==>newvalue
```

And then in lowerTrees (or code generation), if the resolve check was seen to have been applied to a `newvalue`, the helper call would be generated with a `resolveType` of `J9TR_FLAT_RESOLVE_WITHFIELD`. 

One potential, but likely very minor, concern with associating the symbol reference for the `withfield` with the `newvalue` is that the symbol reference would only carry any real meaning in the context of the `ResolveCHK`.  For example, if a `withfield` is applied to the result of another `withfield` operation, we would likely want the JIT to be able to combine the two `newvalue` operations into one - neither symbol reference from the original pair of `newvalue` operations would carry the full meaning for the resultant single `newvalue`, and would need to be dropped in that context.  If either was also a child of a `ResolveCHK`, the original `newvalue` node with its symbol reference would have to be preserved in that context.

**Comment by vijaysun-omr** [comment]
Some comments regarding the prior post in this issue:

You mentioned that "the JIT could generate the following IL, where LHolder.v would be a fake shadow symbol". I am not sure what you mean by a "fake shadow symbol" in this case but I was just picturing it as a normal shadow symbol that gets created for unresolved fields that we have historically had (before we had a notion of value types or flattened fields).  

One other clarification wrt what was shown for the getfield/putfield cases is that I was not picturing the calls to  `jitGetFlattenableField/jitPutFlattenableField` in the IL until the unresolved load/store node themselves get lowered into those calls later in the compilation (should you decide to do so). So the nodes will just look like a load/store as they did all along before we had a notion of value types or flattened fields.

**Comment by hzongaro** [comment]
A further wrinkle I really should have highlighted in the discussion of `withfield` is the (likely, I think) possibility that the class containing the unresolved field might itself be unresolved.  The way `newvalue` currently works, it requires all the fields of the new value that's being created to be supplied.  So in my earlier example,

```
value Point {double x; double y;}
value Line {Point p1; Point p2;}
public static Object sub(Object o, Object v) {
    return (o == null) ? null : (((Line) o) with { p1 = (Point) v; });
}
```

if both `Line.p1` and`Line` are not resolved, the JIT would not be able to generate a `newvalue` to represent that value.  I was assuming that the JIT would use the `jitWithFlattenableField` helper in that case.  The JIT could still create a `ResolveCHK` with a "fake" `newvalue` child to deal with the resolution, but it might look a little odd.

```
      n18n    checkcast 
      n16n      aload  <parm 0 Ljava/lang/Object;>
      n17n      loadaddr  Unres4$Line
      n21n    checkcast 
      n19n      aload  <parm 1 Ljava/lang/Object;>
      n20n      loadaddr  Unres4$Point
      n30n    ResolveCHK [#32]
      n29n      newvalue  LUnres4$Line.p1
      n24n        loadaddr  Unres4$Line
      n16n        ==>aload
      n26n    treetop                                                                              
      n25n      acall  jitWithFlattenableField
      n19n        ==>aload
      n16n        ==>aload
      n24n        aconst 0x186250  
 ```

**Comment by vijaysun-omr** [comment]
Regarding the comment "One potential, but likely very minor, concern with associating the symbol reference for the withfield with the newvalue is that the symbol reference would only carry any real meaning in the context of the ResolveCHK"
I am not sure I completely agree. 

The symbol reference on the newvalue represents the lone field that the "withfield" bytecode (and the "with" syntax in Java) seems to allow modification of given the "original" value type instance. This is true regardless of whether the field in question was resolved or not and I would suggest that it is a property that is relevant for the "newvalue" IL opcode in general, e.g. an optimization may want to know which field was being modified by a given "newvalue" node. 

Regarding "combining" newvalue nodes, that is an interesting scenario. I'm unsure if the "combined" newvalue (that effectively modifies multiple fields as opposed to the "single" newvalue that only modifies a single field) should be a separate IL opcode allowing us to specify multiple sym refs being changed at once. The other alternative is to change the semantics of the newvalue opcode and allow it to refer to multiple sym refs but in either case, I'm unclear on what the issue is that you were concerned about, i.e. if any of the newvalues being combined had a ResolveCHK above it then it would seem that the combined newvalue would need to have a ResolveCHK above it too.

**Comment by Leonardo2718** [comment]
> The newvalue opcode is at least implicitly akin to a store (except that we don't want to expose the store to the optimizer lest it be be separated from the associated new allocation) and so in a sense we could allow a resolve check above it in an analogous manner to what we do for stores.

I'm not sure this is accurate. I think `newvalue` is more akin to a `new` than to a store. `newvalue` represents the instantiation of a new value type instance, not a specific operation on a field.

> Regarding the comment "One potential, but likely very minor, concern with associating the symbol reference for the withfield with the newvalue is that the symbol reference would only carry any real meaning in the context of the ResolveCHK"
> I am not sure I completely agree.
> 
> The symbol reference on the newvalue represents the lone field that the "withfield" bytecode (and the "with" syntax in Java) seems to allow modification of given the "original" value type instance. This is true regardless of whether the field in question was resolved or not and I would suggest that it is a property that is relevant for the "newvalue" IL opcode in general, e.g. an optimization may want to know which field was being modified by a given "newvalue" node.

The issue is that `newvalue` is not specific to `withfield`. Rather, it's `withfield` that is implemented in terms of `newvalue`. `newvalue` is used for other operations as well, such as `defaultvalue` and `getfield`. 

Another concern is that `newvalue` already has a symbol reference associated with it for the VM helper that is called to allocate a value type. Although `newvalue` does get lowered to a `new`, the symref on it is carried over from `newvalue` because the VM helper that is called to do the allocation is not the same one that is used for "regular" `new`s. Eventually, we will want to add support for `newvalue` to the code generators so that optimized allocation sequences can be generated instead of relying on an unconditional helper call.

See https://github.com/eclipse/omr/issues/4028 and https://www.youtube.com/channel/UCO8MkTRrFnLefi5nqReOMMA for a detailed overview of `newvalue`.

> Regarding "combining" newvalue nodes, that is an interesting scenario. I'm unsure if the "combined" newvalue (that effectively modifies multiple fields as opposed to the "single" newvalue that only modifies a single field) should be a separate IL opcode allowing us to specify multiple sym refs being changed at once. The other alternative is to change the semantics of the newvalue opcode and allow it to refer to multiple sym refs but in either case, I'm unclear on what the issue is that you were concerned about, i.e. if any of the newvalues being combined had a ResolveCHK above it then it would seem that the combined newvalue would need to have a ResolveCHK above it too.

This scenario of combining multiple value type operations into a single `newvalue` node was part of the motivation for introducing the `newvalue` opcode as it's expected to provide significant performance benefits.

> For the "flattened" case, I wonder if the same newvalue opcode (with the above proposed changes to use the resolve check above it and to use the symbol reference for the field mentioned in the withfield bytecode) could be used at least through the optimizer and then only lower it differently before codegen into that jitWithFlattenableField helper if you wanted to do so.
>
> This has one other advantage of keeping the IL trees that the optimizer sees relatively consistent in these cases and the differences between flattened and not flattened only occur during lowering.

I'm not sure I follow. For the unresolved cases, we don't know whether the field is flattened, so the unflattened and flattened cases must look the same in the IL. When the fields are resolved, I think we _do_ want the IL to look different so that we can properly optimize it. For example, we will want to be able to scalarize value type instances at some point, since this is one of the "promised" optimization opportunities value types provide. If the optimizer can't see what's been flattened into a type, we can't scalarize them.

> A further wrinkle I really should have highlighted in the discussion of withfield is the (likely, I think) possibility that the class containing the unresolved field might itself be unresolved.

My understanding here is that resolving a field necessarily requires that the type containing the field also be resolved, i.e. if both are unresolved, resolving the field must trigger resolution of the containing type. I believe the existing code checks if the containing type is resolved only as an early exit because we don't handle anything being unresolved. We should definitely keep this case in mind though.

**Comment by hzongaro** [comment]
> You mentioned that "the JIT could generate the following IL, where LHolder.v would be a fake shadow symbol". I am not sure what you mean by a "fake shadow symbol" in this case but I was just picturing it as a normal shadow symbol

The reason I was calling it a fake shadow symbol is that that JIT would create it as if it was to be used to access a reference field in the object that contained it, but in reality, the JVM might have flattened the field as an aggregate into the containing object.  The generated code would never actually attempt to load an address through that symbol, so it wouldn't really matter.  Perhaps I should have referred to it as a "placeholder", to avoid any pejorative connotation.

> One other clarification wrt what was shown for the getfield/putfield cases is that I was not picturing the calls to jitGetFlattenableField/jitPutFlattenableField in the IL until the unresolved load/store node themselves get lowered into those calls later in the compilation (should you decide to do so). So the nodes will just look like a load/store as they did all along before we had a notion of value types or flattened fields.

Yes, that would be simple enough, though I would assume that in cases where the field is already resolved during IL generation the JIT would continue to do as it does today - that is, it would either generate the calls to `jitGetFlattenableField` or `jitPutFlattenableField` during IL generation, or generate IL that operates directly on the contents of the flattened field during IL generation (the choice based on a JIT option for now), rather than using similar placeholder load/store nodes and delaying the choice of how to expand them until code generation as well.

**Comment by vijaysun-omr** [comment]
Regarding "I'm not sure this is accurate. I think newvalue is more akin to a new than to a store. newvalue represents the instantiation of a new value type instance, not a specific operation on a field."

I think the logical equivalent of `newvalue` is a `new` followed by a store. The only reason we don't do it via those existing opcodes is that we don't want the optimizer to separate those 2 IL trees, correct ? If you agree with these statements, I'm not sure why it matters if `'newvalue` is more akin to a `new` or to a store since it does both. 
The fact that a resolution can occur at the `newvalue` for the field being stored means that that is worth capturing to honestly reflect what was in the bytecodes in the IL, i.e. there was a constant pool reference for the field being written to in the `withfield` and I thought it was relatively uncontroversial that we ought to be including that constant pool information (via a symbol reference as we usually do in our IL) at a `newvalue`.

**Comment by Leonardo2718** [comment]
After an offline discussion to clear up a number of background details on the topic, @vijaysun-omr and I have come up with a concrete decision to move this issue forward.

## Background

The core problem we are trying to solve is to find a place to put the information needed to implement field resolution. We, therefore, must consider what information is needed by the different parts of the compiler.

From a functional point of view, the optimizer only needs to know that an exception may be thrown at the point where a resolution may occur. `ResolveCHK` nodes serve this purpose by indicating an operation may throw an exception (due to resolution) and _do not_ serve any additional purpose. As a result, a `ResolveCHK` doesn't really add value to operations that may already throw exceptions. For example, a call that can throw any exception does not need a `ResolveCHK` even though it may internally trigger the resolution of something.

On the other hand, code generators need to have access to all information needed to generate the resolution code (dispatch to unresolved data snippet, etc).

## Proposal

Value type field resolution will be done by a call to the `jitResolveFlattenableField` VM helper. So, all that the code generators need is a symbol for the helper and the arguments to pass to the call.

Since a call (that can throw anything) is all that's needed for the optimizer to functionally handle resolution (i.e. not touching it), and resolution is done by a call, simply having a call to the helper in the IL is enough to represent resolution. Using a non-helper symbol, as was suggested in the first alternative presented at the beginning of this issue, would also work.

```
    n99n   treetop
    n98n     acall valueTypeFieldResolveCHK
    n97n       iconst 5  // cpIndex
    n96n       iconst 1  // J9TR_FLAT_RESOLVE_GETFIELD
    n21n   treetop
    n20n     acall  jitGetFlattenableField
    n16n       aload  <parm 0 LHolder;>
    n17n       aconst 0x17e210   // cp()+cpIndex
```

In either case, special handling for the call will likely be needed in the code generators. In addition, as long as the symbol is specified to throw any exception, the optimizer will behave correctly.

As an alternative, a non-helper symbol could be defined to represent the combined resolution and field operation. However, there is no clear benefit to this approach and it may require a little more work because of the need to reimplement the field operation (in lowerTrees or the code gens).

## Unanswered questions

- It's still unclear whether a non-helper symbol should be used or whether we can just use the symbol for the VM helper.
- We also need to clarify what exactly the code generators are going to do/generate.

## Considerations

During our discussion, there were a number of points that came up that I believe are worth mentioning for completeness.

- It's important to remember that `newvalue` can't be used after the resolution helper call because the layout of the field's type is unknown.
- A key motivation for the design of `newvalue` is the ability to unlock optimizations like choosing when to allocate new instances (instead of mutating an existing one) and scalarizing instances.
- One benefit of `ResulveCHK` nodes is that the optimizer can eliminate unneeded occurrences, for example, when one instance is dominated by an equivalent instance. However, @vijaysun-omr noted the benefit of this optimization is minimal and unlikely to be significant.
- Using load nodes for `getfield` would be problematic because a getfield of a value type is not a "true" load. It has additional semantics that are not captured by the existing load opcodes and would cause problems for the optimizer. For example, `getfield` may result in an allocation (because of boxing), which implies that the operation is a GC point and may throw an exception. However, the optimizer expects loads to never throw. A similar argument can be made for `withfield` and stores.
- @vijaysun-omr believes paths with unresolved fields are likely not worth optimizing. Specifically, it may not be worth trying to patch the resolution code.
- For `withfield`, we considered the possibility of using a `newvalue` where the field being "changed" would have a special symref representing the entire field, including all flattened sub-fields. Doing so, however, would overload the semantics of `newvalue` and complicate future uses of the opcode in optimizations.
- For `withfield`, although it should be possible to set the symref of the field being "changed" on the `newvalue` node, doing so seems a little inconsistent because there is nothing about `newvalue` that ties it to a specific field or operation. Other than field resolution, we don't generally care about _where_ a value type comes from (i.e. `defaultvalue` vs `getfield` vs `withfield`). We only care about what we can do with the resulting instance, which is represented by a `newvalue` node. Therefore, it's reasonable for `newvalue` and its properties to _not_ depend on what operation it came from.
- An important benefit of using a consistent, `newvalue`-based representation for value type operations is that any optimizations targeting value types only need to look at one opcode/representation (i.e. optimizations don't have to be taught how to deal with different representations for `defaultvalue`, `gefield`, and so on, individually).

**Comment by 0xdaryl** [comment]
Thanks for summarizing the discussion Leo.

A couple questions around this example:
```
    n99n   treetop
    n98n     acall valueTypeFieldResolveCHK
    n97n       iconst 5  // cpIndex
    n96n       iconst 1  // J9TR_FLAT_RESOLVE_GETFIELD
```
1)  Is `valueTypeFieldResolveCHK` part of your "alternate non-helper with combined resolution and field operation" example?  Otherwise, why wouldn't this simply be a call to the `jitResolveFlattenableField` helper ?

2)  I think the API needs a RAM Method or a constant pool address to go along with the cpIndex.  If not, where is the cpAddr coming from?

**Comment by fjeremic** [comment]
Just dropping in to say I appreciate that offline discussions are being summarized here, or the discussions are held here exclusively. This helps those of us on the sidelines understand what is motivating the work that is to come. Thanks again and keep it up!

**Comment by hzongaro** [comment]
In prototyping JIT support for calling the new resolution helper, we realized the JIT would need an additional helper from the JVM to handle `withfield`.

For a `withfield` operation on a field of a type that has been resolved, the JIT generates a call to the `jitNewValue` helper to create a new instance, copy the fields from the old value, and update the value of the single field that is the subject of the `withfield` operation.

If the type is unresolved, the JIT has no way of knowing what fields to copy from the old value to the new value.  In pull request #12094, the JVM provides a new `jitCloneValueType`.  The JIT can generate a call to that helper to clone a value type, then update the one field that's the subject of the `withfield` operation using IL similar to that used for a `putfield` that stores into an unresolved field.

**Comment by hzongaro** [comment]
Adding a very belated update to this discussion.  As `withfield` is no longer a bytecode operation in the prototype implementation of value types, the portions of this discussion that relate to that operation are no longer relevant.
